tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/Linker.scala

```diff
@@ -38,50 +38,51 @@ final class Linker(semantics: Semantics, outputMode: OutputMode,
   def this(semantics: Semantics, considerPositions: Boolean) =
     this(semantics, OutputMode.ECMAScript51Isolated, considerPositions)
 
-  private[this] val files = mutable.Map.empty[String, PersistentIRFile]
-
-  private[this] var statsReused: Int = 0
-  private[this] var statsInvalidated: Int = 0
-  private[this] var statsTreesRead: Int = 0
-
   type TreeProvider = String => (ClassDef, Option[String])
 
-  /** Cleans the cache. */
-  def clean(): Unit = {
-    files.clear()
-    statsReused = 0
-    statsInvalidated = 0
-    statsTreesRead = 0
+  /** Obsolete method. Does not do anything. */
+  @deprecated("Does not do anything. Don't use", "0.6.5")
+  def clean(): Unit = ()
+
+  // Binary compat
+  def link(irInput: immutable.Traversable[VirtualScalaJSIRFile], logger: Logger,
+      reachOptimizerSymbols: Boolean, bypassLinkingErrors: Boolean,
+      noWarnMissing: immutable.Seq[NoWarnMissing],
+      @deprecatedName('checkInfos) checkIR: Boolean): LinkingUnit = {
+    link(irInput: Traversable[VirtualScalaJSIRFile], logger,
+      reachOptimizerSymbols, bypassLinkingErrors, noWarnMissing, checkIR)
   }
 
   def link(irInput: Traversable[VirtualScalaJSIRFile], logger: Logger,
       reachOptimizerSymbols: Boolean, bypassLinkingErrors: Boolean,
-      noWarnMissing: Seq[NoWarnMissing],
+      noWarnMissing: immutable.Seq[NoWarnMissing],
       @deprecatedName('checkInfos) checkIR: Boolean): LinkingUnit = {
-    startRun()
 
-    val encodedNameToPersistentFile =
-      logTime(logger, "Linker: Read info")(updateFiles(irInput))
+    val infosBuilder = List.newBuilder[Infos.ClassInfo]
+    val encodedNameToFileBuilder = Map.newBuilder[String, VirtualScalaJSIRFile]
 
-    val infos = encodedNameToPersistentFile.values.map(_.info).toList
+    for (irFile <- irInput) {
+      val info = irFile.info
+      infosBuilder += info
+      encodedNameToFileBuilder += info.encodedName -> irFile
+    }
+
+    val infos = infosBuilder.result
+    val encodedNameToFile = encodedNameToFileBuilder.result
```

Comment:
`result()` with `()` (twice). Calling `result()` has side effects, in theory, since it invalidates the state of the builder.